File: TimeFre_analysis/Hilbert/hilbert_analysis.py
import numpy as np
from scipy.signal import butter, filtfilt, hilbert

def hilbert_analysis(data, srate, frequency_bands):
    """
    Perform Hilbert Transform for power and phase extraction.

    Parameters:
    -----------
    data : array-like, (n_samples, n_trials)
    srate : float
    frequency_bands : list of tuples/lists, e.g. [[4, 8], [8, 12]]

    Returns:
    --------
    tf_power : array, (n_bands, n_samples, n_trials)
    tf_phase : array, (n_bands, n_samples, n_trials)
    """
    n_samples, n_trials = data.shape
    n_bands = len(frequency_bands)

    tf_power = np.zeros((n_bands, n_samples, n_trials))
    tf_phase = np.zeros((n_bands, n_samples, n_trials))

    nyquist = srate / 2.0

    for b_idx, band in enumerate(frequency_bands):
        low, high = band
        # Design filter
        b, a = butter(4, [low / nyquist, high / nyquist], btype='bandpass')

        for i in range(n_trials):
            # Filter
            filtered = filtfilt(b, a, data[:, i])

            # Hilbert
            analytic = hilbert(filtered)

            # Extract features
            tf_power[b_idx, :, i] = np.abs(analytic)**2
            tf_phase[b_idx, :, i] = np.angle(analytic)

    return tf_power, tf_phase

File: TimeFre_analysis/Hilbert/hilbert_analysis.m
function [tf, phase] = hilbert_analysis(data, srate, frequency_bands)
%HILBERT_ANALYSIS Hilbert transform for instantaneous power and phase
%   [tf, phase] = hilbert_analysis(data, srate, frequency_bands)
%
%   Inputs:
%       data:            Signal data (samples x trials)
%       srate:           Sampling rate (Hz)
%       frequency_bands: Cell array of frequency bands, e.g., {[4 8], [8 12]}
%                        or matrix [min1 max1; min2 max2]
%
%   Outputs:
%       tf:              Power envelope (bands x samples x trials)
%       phase:           Instantaneous phase (bands x samples x trials)

    n_samples = size(data, 1);
    n_trials = size(data, 2);

    if iscell(frequency_bands)
        n_bands = length(frequency_bands);
        bands = zeros(n_bands, 2);
        for i=1:n_bands
            bands(i,:) = frequency_bands{i};
        end
    else
        n_bands = size(frequency_bands, 1);
        bands = frequency_bands;
    end

    tf = zeros(n_bands, n_samples, n_trials);
    phase = zeros(n_bands, n_samples, n_trials);

    for b = 1:n_bands
        f_low = bands(b, 1);
        f_high = bands(b, 2);

        % Design filter (simple Butterworth bandpass)
        nyquist = srate / 2;
        [b_filt, a_filt] = butter(4, [f_low, f_high] / nyquist, 'bandpass');

        for trial = 1:n_trials
            % Filter data
            filt_data = filtfilt(b_filt, a_filt, data(:, trial));

            % Hilbert transform
            analytic_signal = hilbert(filt_data);

            % Power (envelope squared)
            tf(b, :, trial) = abs(analytic_signal).^2;

            % Phase
            phase(b, :, trial) = angle(analytic_signal);
        end
    end

end
